### PR TITLE
more changes for building with vs2019/msvc

### PR DESCRIPTION
**Pull request by greg7mdp**

### CODE REVIEW

File: filament/src/components/LightManager.h [hunk 1]
```diff
@@ -244,7 +244,7 @@ class FLightManager : public LightManager {
         struct Proxy {
             // all of this gets inlined
             UTILS_ALWAYS_INLINE
-            constexpr Proxy(Base& sim, utils::EntityInstanceBase::Type i) noexcept
+            /* constexpr */ Proxy(Base& sim, utils::EntityInstanceBase::Type i) noexcept
```
Comment (romainguy) [inline filament/src/components/LightManager.h]: If `constexpr` isn't correct, just remove it, don't comment it. Same for the instances below.

File: third_party/robin-map/tsl/robin_hash.h [hunk 2]
```diff
@@ -1116,11 +1116,26 @@ class robin_hash: private Hash, private KeyEqual, private GrowthPolicy {
     void insert_value(std::size_t ibucket, distance_type dist_from_ideal_bucket, 
                       truncated_hash_type hash, Args&&... value_type_args) 
     {
-        insert_value(ibucket, dist_from_ideal_bucket, hash, value_type(std::forward<Args>(value_type_args)...));
+        value_type value(std::forward<Args>(value_type_args)...);
```
Comment (romainguy) [inline third_party/robin-map/tsl/robin_hash.h]: Is this an update of robin-map from upstream or changes you made yourself? If the latter, we should upstream those changes or at least track them in a README somewhere otherwise we'll lose them.

Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: This is aa fix that was made in the robin_map github directory. I added the url of the change in a comment.

Comment (romainguy) [reply third_party/robin-map/tsl/robin_hash.h]: We should probably update `robin-map` entirely instead of taking single changes like this. I believe we have a .txt in the robin-map directory that explains how to do this.

File: filament/src/components/LightManager.h (hunk 1 above)
Comment (greg7mdp) [reply filament/src/components/LightManager.h]: @romainguy, done.

File: third_party/robin-map/tsl/robin_hash.h (hunk 2 above)
Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: Well, the change is pretty specific got a msvc issue and is not dependent on other changes. I was thinking that if you guys want to pull the latest version of robin-map, it would probably be better for you guys to do it. That being said I can do it if you wish (I don't see the .txt file you refer to in robin-map directory).

Comment (bejado) [reply third_party/robin-map/tsl/robin_hash.h]: I'll update `robin-map` to the latest master

Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: great, thanks @bejado. I'll remove the robin-map change from my PR then.

Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: @bejado, I did remove my change in `robin-map`.

Comment (bejado) [reply third_party/robin-map/tsl/robin_hash.h]: So the latest `robin-map` does not appear to be compatible with Visual Studio 2015 headers. It works with VS 2017, but given that our CI only has 2015, I think we should proceed with this single cherry-picked change. @romainguy what do you think?

Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: I'll be happy to add it back, I agree it seems like the best solution at this point.

Comment (greg7mdp) [reply third_party/robin-map/tsl/robin_hash.h]: @bejado, I added the cherry-picked change to `robin-map` back

**Comment by greg7mdp** [comment]
@romainguy, are we just waiting for another review on this?

**Comment by romainguy** [comment]
I'd like @pixelflinger and @bejado to review these changes. Also, I'm on vacation this week :0

**Comment by greg7mdp** [comment]
OK, sorry for being pushy @romainguy, even on vacation you answer immediately :-)! Bonnes vacances!

**Comment by bejado** [comment]
Looks like our CI is failing (unrelated to these changes), so I'd like to investigate and get those green before merging. Otherwise LGTM.

File: filament/backend/include/private/backend/CommandStream.h [hunk 3]
```diff
@@ -145,8 +145,20 @@ struct CommandType<void (Driver::*)(ARGS...)> {
...
-        template<std::size_t... I> void log(std::index_sequence<I...>) noexcept;
+        // these two log() template functions should be defined inline in this class 
+        // definition, otherwise when building with msvc they are undefined. Alternatively they 
+        // should be defined in a file which calls them, not CommandStream.cpp.
+        void log() noexcept { log(std::make_index_sequence<std::tuple_size<SavedParameters>::value>{}); }
+
+        template<std::size_t... I> void log(std::index_sequence<I...>) noexcept {
+#if DEBUG_COMMAND_STREAM
+            static_assert(UTILS_HAS_RTTI, "DEBUG_COMMAND_STREAM can only be used with RTTI");
+            std::string command = utils::CallStack::demangleTypeName(typeid(Command).name()).c_str();
```
Comment (pixelflinger) [inline filament/backend/include/private/backend/CommandStream.h]: uhm I think there is a problem here. Now the header file depends on <string>, and we don't want headers to include <string>. I'm not sure why having the definition in the .cpp doesn't work, since that's the only place it's used from?

Oh... I think maybe all you need is to surround them with `#if DEBUG_COMMAND_STREAM` in the cpp (or hide the declaration when DEBUG_COMMAND_STREAM is not used.

Comment (greg7mdp) [reply filament/backend/include/private/backend/CommandStream.h]: @pixelflinger I'll try this

Comment (greg7mdp) [reply filament/backend/include/private/backend/CommandStream.h]: @pixelflinger thanks for pointing me to the right direction there was a call to log() protected by a if statement instead of an #ifdef. I have updated the PR.

Comment (greg7mdp) [reply filament/backend/include/private/backend/CommandStream.h]: @pixelflinger are you OK with the change?

**Comment by greg7mdp** [comment]
@bejado, is the CI issue fixed?

**Comment by bejado** [comment]
CI issue should be fixed now. I'll defer to @pixelflinger for the final LGTM.

**Comment by greg7mdp** [comment]
@pixelflinger, are you OK with the changes?

**Comment by greg7mdp** [comment]
@pixelflinger, are you OK with the changes?

**Comment by romainguy** [comment]
Now @pixelflinger is on vacation for the week :) Sorry about that, we've be able to give your PRs a lot more attention starting next week. I really appreciate both your hard work and patience, we're really grateful for these changes.

**Comment by greg7mdp** [comment]
OK, many thanks @romainguy for the update, I will be patient!

**Comment by romainguy** [comment]
Note: we're moving our CI & presubmit builds from our internal system to GitHub Actions. This will allow us to build Filament with Visual Studio 2019 easily.

**Comment by greg7mdp** [comment]
This is great news, thanks @romainguy . I hope we can accelerate merging my changes somewhat, so the master branch can actually build with MSVC frrom vs2019 (like it does on my side). Once we have the CI checking native MSVC builds it will guarantee these are not broken, which will be great!

**Comment by greg7mdp** [comment]
@pixelflinger when do you think you can review this PR?

**Comment by bejado** [comment]
Hey @greg7mdp, do you have additional changes (other than this PR) that need to be merged to support MSVC 2019? I'm trying to build this branch, but still running into a lot of issues even with the latest MSVC.

**Comment by greg7mdp** [comment]
No, there are other changes. I was trying to submit one PR at a time so I wouldn't overwhelm the filament devs, but I couldn't get this last PR reviewed and it got stuck.
If you want I could submit a PR with all my changes to build with vs2019, Many of the changes are in various CMakeLists.txt.

**Comment by greg7mdp** [comment]
@bejado, after waiting over a month for my PR to be reviewed, now I do have a number of merge conflicts to resolve, so it is going to take me some time to resolve them. I'll let you know when I'm done.

File: filament/backend/include/private/backend/CommandStream.h [hunk 4]
```diff
@@ -153,10 +153,10 @@ struct CommandType<void (Driver::*)(ARGS...)> {
         static inline void execute(M&& method, D&& driver, CommandBase* base, intptr_t* next) noexcept {
             Command* self = static_cast<Command*>(base);
             *next = align(sizeof(Command));
-            if (DEBUG_COMMAND_STREAM) {
+#if DEBUG_COMMAND_STREAM
```
Comment (pixelflinger) [inline filament/backend/include/private/backend/CommandStream.h]: Why this change here?

Comment (greg7mdp) [reply filament/backend/include/private/backend/CommandStream.h]: Because when you do:

```
if (false) 
    self->log();
```

the MSVC compiler, when building in debug mode, generates a call to self->log() in the executable and we get an undefined symbol, because the function is defined in CommandStream.cpp with:


```
#if DEBUG_COMMAND_STREAM
#define DECL_DRIVER_API_SYNCHRONOUS(RetType, methodName, paramsDecl, params)
#define DECL_DRIVER_API(methodName, paramsDecl, params) \
    template void CommandType<decltype(&Driver::methodName)>::Command<&Driver::methodName>::log() noexcept;
#define DECL_DRIVER_API_RETURN(RetType, methodName, paramsDecl, params) \
    template void CommandType<decltype(&Driver::methodName##R)>::Command<&Driver::methodName##R>::log() noexcept;
#include "private/backend/DriverAPI.inc"
#endif
```

The change just uses the preprocessor (#if) in both `CommandStream.h` and `CommandStream.cpp` so it is consistent, instead if using if in the .h and #if ion the .cpp.

**Comment by bejado** [comment]
@greg7mdp Sorry about the delay. After the merge conflicts are fixed, this PR should be good to merge. As for the rest of the changes, feel free to put them in a single PR (you can use multiple commits to separate logical changes if you wish). That way we can review them all together.

**Comment by greg7mdp** [comment]
@bejado, unfortunately I am having a very hard time building filament with MSVC after grabbing the latest changes. Mathias changed the DriverAPI.inc macros in a way that doesn't build anymore with MSVC. I'm getting many errors like:

```
1>C:\greg\github\filament\filament\backend\include\private/backend/DriverAPI.inc(106,1): error C2061:  syntax error: identifier 'PAIR_ARGS_E'
1>C:\greg\github\filament\filament\backend\include\private/backend/DriverAPI.inc(109,1): error C2061:  syntax error: identifier 'PAIR_ARGS_E'
1>C:\greg\github\filament\filament\backend\include\private/backend/DriverAPI.inc(112,1): error C2061:  syntax error: identifier 'PAIR_ARGS_E'
```

I tried including the older version of the file instead, using:

```
#if defined(_MSC_VER)
    #include "./DriverAPI.inc.msvc"
#else
...
```

But then I have error when building filament:

```
4>Generating venetian_crossroads_2k/venetian_crossroads_2k_skybox.ktx, venetian_crossroads_2k/venetian_crossroads_2k_ibl.ktx
1>OpenGLDriver.cpp
1>C:\greg\github\filament\filament\backend\include\private\backend\./DriverAPI.inc.msvc(405,1): error C2672:  'filament::backend::CommandType<void (__cdecl filament::backend::Driver::* )(int32_t,int32_t,uint32_t,uint32_t)>::Command<void filament::backend::Driver::setViewportScissor(int32_t,int32_t,uint32_t,uint32_t)>::execute': no matching overloaded function found
1>C:\greg\github\filament\filament\backend\include\private\backend\./DriverAPI.inc.msvc(405): message :  while compiling class template member function 'void filament::backend::ConcreteDispatcher<filament::OpenGLDriver>::setViewportScissor(filament::backend::Driver &,filament::backend::CommandBase *,intptr_t *)'
1>C:\greg\github\filament\filament\backend\src\opengl\OpenGLDriver.cpp(132): message :  see reference to class template instantiation 'filament::backend::ConcreteDispatcher<filament::OpenGLDriver>' being compiled
1>C:\greg\github\filament\third_party\robin-map\tsl\robin_growth_policy.h(206): message :  see reference to class template instantiation 'std::array<size_t,39>' being compiled
1>C:\greg\github\filament\filament\backend\include\private\backend\./DriverAPI.inc.msvc(409,1): error C2783:  'void filament::backend::CommandType<void (__cdecl filament::backend::Driver::* )(int32_t,int32_t,uint32_t,uint32_t)>::Command<void filament::backend::Driver::setViewportScissor(int32_t,int32_t,uint32_t,uint32_t)>::execute(M &&,D &&,filament::backend::CommandBase *,intptr_t *) noexcept': could not deduce template argument for 'M'
1>C:\greg\github\filament\filament\backend\include\private/backend/CommandStream.h(153): message :  see declaration of 'filament::backend::CommandType<void (__cdecl filament::backend::Driver::* )(int32_t,int32_t,uint32_t,uint32_t)>::Command<void filament::backend::Driver::setViewportScissor(int32_t,int32_t,uint32_t,uint32_t)>::execute'
```

At this point I don't really know what to do. I can commit my current state, but then all my changes will be added to this PR I think. I think most of the changes are safe, and I guess you guys could pick and choose which ones you want.

Or I could work on getting filament to build again with MSVC first, but it will take me a few days probably.

**Comment by greg7mdp** [comment]
I tried to create a new branch (greg7mdp/cmake) to submit my other changes separately, but I get this error in github:

image

Do you know what the problem might be? I don't know git well at all.

**Comment by bejado** [comment]
@greg7mdp I haven't used the GitHub desktop client, but looks like the relevant part of the error is:

> refusing to allow an integration to create or update .github/workflows/presubmit.yml

[This issue](https://github.com/desktop/desktop/issues/7079) seems related. Maybe try removing that file on your side? It's unrelated to MSVC changes.

At this point what I'd recommend is pushing all your MSVC changes to a branch and creating a PR, even if it's out-of-date with our master branch. Then we can look at the merge conflicts and help decide how to best fix them.

**Comment by greg7mdp** [comment]
@bejado, @romainguy, I caught up this branch, I think it is ready to merge now.

**Comment by bejado** [comment]
CI is failing, but unrelated to these changes :/ I'm able to build locally, so I'll go ahead and merge.

**Comment by greg7mdp** [comment]
Thanks @bejado!
